![Blue Bank](https://finch-groundhog-9245.squarespace.com/s/Blue-Bank-Logo.png "Blue Bank")

> **Project Title:** Loan Analysis<br>
> **Project Owner:** Berlinda Anaman <br>
> **Email:** Berlana.d@gmail.com <br>
> **Github Profile:** https://github.com/berl-cloud<br>
> **LinkedIn Profile:** http://www.linkedin.com/in/berlinda-anaman<br>
> **Tableau Profile:** https://public.tableau.com/app/profile/berlinda.anaman

## Table of Contents<a id='mu'></a>

* [Business Problem Understanding](#bpu)
    * [Problem Statement](#ps)
    * [Project Goal](#pg)
    * [Information Needed](#in)
    * [Methodology](#my)
* [Data Preparation](#dp)
    * [Data Understanding](#du)
    * [Data Quality Assessment](#dqa)
       * [Loan Data](#ld)
    * [Data Cleaning and Preprocessing](#dcp)
* [Data Visualisation](#dv)
* [References](#r)

## 1. Business Problem Understanding<a id='bpu'></a>

[Move Up](#mu)

<p style="text-align:justify;">The first step in approaching a data science problem is problem understanding. This step is very important since it allows us to know the kind of decisions we want to make, the information or data that will be needed to inform those decisions and finally, the kind of analysis that will be used to arrive at those decisions. In a nutshell, developing a mental model of the problem allows us to properly structure potentially relevant information needed to solve the problem.</p>

### 1.1 Problem Statement <a id='ps'></a>

Blue Bank, a US-based bank, is facing challenges in its loan department due to being understaffed. The bank supplies loans to individuals but lacks sufficient reporting on the riskiness of borrowers. To mitigate potential losses and improve decision-making, the bank aims to conduct a loan risk analysis to identify borrowers who may have difficulties paying back their loans. The analysis will be carried out using Python and visualized using Tableau to create a comprehensive and actionable report.

### 1.2 Project Goal <a id='pg'></a>

In this project,our main goal is to create a report on the riskiness of loan borrowers. We will build a Tableau dashboard to measure the following;

* Average days with Credit Line.
* Maximum Revolving Balance.
* Maximum and Minimum Interest rates.
* Average Debt to Income Ratio.
* Total Number of Borrowers with the highest and lowest interest rates.
* Total Number of Borrowers with the highest utilization rate. 
* Annual income range of borrowers.
* Total Number of Borrowers with good or bad FICO Credit Scores.

### 1.3 Information Needed <a id='in'></a>

To complete the loan risk analysis and create a comprehensive report, the following data and information will be required:

1. **Borrower Data:** A dataset containing historical information on borrowers, including credit scores, income, existing debts, loan amounts, purpose of loan, dept to income ratio, and loan status (e.g., fully paid, delinquent, default, etc.).

2. **Loan Performance Data:** Data on the performance of past loans, including repayment behavior, delinquency rates, default rates, and any relevant indicators of loan success or failure.

3. **Tableau Visualizations:** Tableau workbooks or interactive visualizations that present loan risk insights in a clear and intuitive manner.

4. **Project Execution Steps:** A detailed outline of the steps taken to perform the loan risk analysis, including data preprocessing and visualization design.

### 1.4 Methodology <a id='my'></a>

<p style="text-align:justify;">The methodology that will be used for our project will largely depend on the goals we set out to achieve. The methodlogy framework below gives us a comprehensive guide on the methodology apparoach that will help us achieve our goals.</p>
<br>
<p style="text-align:center;font-weight:bold;font-size:20px"> Methodology Framework</p>
<br>
<img src='https://artofdatablog.files.wordpress.com/2017/10/methodology-map.jpg' style="float:center;width:700px;">

Once we have the data, we would need to understand it, clean it and perform the analysis to generate insights.

## 2. Data Preparation <a id='dp'></a>

[Move Up](#mu)

An understanding of the data coupled with problem understanding will help us in cleaning and preparing our data for analysis. It is usually rare to acquire a ready-to-use data for any analysis without some level of preparation. To prepare our data, we normally assess the quality of the data, cleanse, format, blend and sample the data since we may encounter various issues with columns in our data. These issues may include:

* **`Missing values:`** meaning column values are incomplete
* **`Incorrect data:`** meaning you see values not expected for the column name
* **`Inconsistent values:`** meaning some values may fall outside the expected range
* **`Duplicate values:`** meaning whether or not there are duplicate values
* **`Inconsistent data type:`** meaning values entered in the columns may not be consistent with the column names

To properly prepare our data for analysis, we will perform two important tasks which are;

* Part I: Data Quality Assessment
* Part II: Data Cleaning and Preprocessing 

### 2.1 Data Understanding <a id='du'></a>

**Description of Data Columns**

Below are the various columns of the loan data and what they represent:

* `credit.policy:` 1 if the customer meets the credit underwriting criteria of Blue Bank, and 0
otherwise.
* `purpose:` The purpose of the loan (takes values "credit_card", "debt_consolidation",
"educational", "major_purchase", "small_business", and "all_other").
* `int.rate:` The interest rate of the loan, as a proportion (a rate of 11% would be stored as 0.11).
Borrowers judged by Blue Bank to be more risky are assigned higher interest rates.
* `installment:` The monthly installments owed by the borrower if the loan is funded.
* `log.annual.inc:` The natural log of the self-reported annual income of the borrower.
* `dti:` The debt-to-income ratio of the borrower (amount of debt divided by annual income).
    - `dti>1` then borrower has more debt than income.
    - `dti<1` then borrower has more income than 
* `fico:` The FICO credit score of the borrower.
    - 300 - 400: Very Poor
    - 401 - 600: Poor
    - 601 - 660: Fair
    - 661 - 780: Good
    - 781 - 850: Excellent
* `days.with.cr.line:` The number of days the borrower has had a credit line.
* `revol.bal:` The borrower's revolving balance (amount unpaid at the end of the credit card
billing cycle).
* `revol.util:` The borrower's revolving line utilization rate (the amount of the credit line used
relative to total credit available).
* `inq.last.6mths:` The borrower's number of inquiries by creditors in the last 6 months. (If there
are a lot of inquiries, that’s an issue)
* `delinq.2yrs:` The number of times the borrower had been 30+ days past due on a payment in
the past 2 years.
* `pub.rec:` The borrower's number of derogatory public records (bankruptcy filings, tax liens,
or judgments).

### 2.2 Data Quality Assessment <a id='dqa'></a>

<p style="text-align:justify;">The first task that we will perform under the data preparation step is initial assessment of the quality of data which will easily allow us to properly clean our data. We will use this section to write any code necesary for inspecting the dataset. Once completed, we will leave our report in the Data Quality Report Document.

At the end of our inspection, we will provide a summary of all of our findings.</p>

In [1]:
# import libraries needed

import json
import pandas as pd
import numpy as np
import warnings


warnings.filterwarnings('ignore')



In [2]:
# load json data

json_file = open("..\Data\loan_data_json.json")
data = json.load(json_file)

In [3]:
# transform to dataframe

loan_data = pd.DataFrame(data)

#### 2.2.1 Inspecting Loan Data <a id='ld'></a>

In this step we will inspect the loan data to get an overview of what the data entails and also assess the quality of the data and make all the possible recommendation for cleaning this data.

In [4]:
# have a glimpse of the data

loan_data.head()

credit.policy             purpose  int.rate  installment  log.annual.inc  \
0              1  debt_consolidation    0.1189       829.10       11.350407   
1              1         credit_card    0.1071       228.22       11.082143   
2              1  debt_consolidation    0.1357       366.86       10.373491   
3              1  debt_consolidation    0.1008       162.34       11.350407   
4              1         credit_card    0.1426       102.92       11.299732   

     dti  fico  days.with.cr.line  revol.bal  revol.util  inq.last.6mths  \
0  19.48   737        5639.958333      28854        52.1               0   
1  14.29   707        2760.000000      33623        76.7               0   
2  11.63   682        4710.000000       3511        25.6               1   
3   8.10   712        2699.958333      33667        73.2               1   
4  14.97   667        4066.000000       4740        39.5               0   

   delinq.2yrs  pub.rec  not.fully.paid  
0            0        0               0  
1            0        0               0  
2            0        0               0  
3            0        0               0  
4            1        0               0

In [5]:
#checking the shape of the data

loan_data.shape

(9578, 14)

> We can see from the above result that the data has `9578` observations and `14` columns. The data is enough to help us perform our analysis. However, we will have to assess the quality of the data and make the necessary cleaning before setting out to achieve our goals.

In [6]:
# checkng data info

loan_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9578 entries, 0 to 9577
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   credit.policy      9578 non-null   int64  
 1   purpose            9578 non-null   object 
 2   int.rate           9578 non-null   float64
 3   installment        9578 non-null   float64
 4   log.annual.inc     9578 non-null   float64
 5   dti                9578 non-null   float64
 6   fico               9578 non-null   int64  
 7   days.with.cr.line  9578 non-null   float64
 8   revol.bal          9578 non-null   int64  
 9   revol.util         9578 non-null   float64
 10  inq.last.6mths     9578 non-null   int64  
 11  delinq.2yrs        9578 non-null   int64  
 12  pub.rec            9578 non-null   int64  
 13  not.fully.paid     9578 non-null   int64  
dtypes: float64(6), int64(7), object(1)
memory usage: 1.0+ MB


In [7]:
# finding unique values for the purpose column

loan_data['purpose'].unique()

array(['debt_consolidation', 'credit_card', 'all_other',
       'home_improvement', 'small_business', 'major_purchase',
       'educational'], dtype=object)

In [8]:
# checking for duplicate values

loan_data.duplicated().sum()

0

In [9]:
 # viewing summary statistics
    
loan_data.describe()

credit.policy     int.rate  installment  log.annual.inc          dti  \
count    9578.000000  9578.000000  9578.000000     9578.000000  9578.000000   
mean        0.804970     0.122640   319.089413       10.932117    12.606679   
std         0.396245     0.026847   207.071301        0.614813     6.883970   
min         0.000000     0.060000    15.670000        7.547502     0.000000   
25%         1.000000     0.103900   163.770000       10.558414     7.212500   
50%         1.000000     0.122100   268.950000       10.928884    12.665000   
75%         1.000000     0.140700   432.762500       11.291293    17.950000   
max         1.000000     0.216400   940.140000       14.528354    29.960000   

              fico  days.with.cr.line     revol.bal   revol.util  \
count  9578.000000        9578.000000  9.578000e+03  9578.000000   
mean    710.846314        4560.767197  1.691396e+04    46.799236   
std      37.970537        2496.930377  3.375619e+04    29.014417   
min     612.000000         178.958333  0.000000e+00     0.000000   
25%     682.000000        2820.000000  3.187000e+03    22.600000   
50%     707.000000        4139.958333  8.596000e+03    46.300000   
75%     737.000000        5730.000000  1.824950e+04    70.900000   
max     827.000000       17639.958330  1.207359e+06   119.000000   

       inq.last.6mths  delinq.2yrs      pub.rec  not.fully.paid  
count     9578.000000  9578.000000  9578.000000     9578.000000  
mean         1.577469     0.163708     0.062122        0.160054  
std          2.200245     0.546215     0.262126        0.366676  
min          0.000000     0.000000     0.000000        0.000000  
25%          0.000000     0.000000     0.000000        0.000000  
50%          1.000000     0.000000     0.000000        0.000000  
75%          2.000000     0.000000     0.000000        0.000000  
max         33.000000    13.000000     5.000000        1.000000

We will proceed by checking if the data contain any missing values. We can easily tell from the results of the info but we want to be extremely sure there are no missing values.

In [10]:
loan_data.isnull().sum()

credit.policy        0
purpose              0
int.rate             0
installment          0
log.annual.inc       0
dti                  0
fico                 0
days.with.cr.line    0
revol.bal            0
revol.util           0
inq.last.6mths       0
delinq.2yrs          0
pub.rec              0
not.fully.paid       0
dtype: int64

#### Data Quality Summary:

This data provides no opportunity for data cleaning. Based on the ouputs displayed above, here is the summary of the data quality assesment:<br>

* It is clear that the data has no missing values, no duplicates and no inconsistencies.


* We will need to convert values of the `log.annual.inc` column to get the actual annual income of the borrowers. We will use the exp() function in doing this.


* Finally, we also need to add some new columns to the loan data:
    * `fico.category`  (determines how good or poor the borrower's credit score is)
    * `int.rate.type` (determines whether the borrower's interest rate is high or low)

### 2.3 Data Cleaning and Preprocessing<a id='dcp'></a>

####  Data Cleaning

<p style="text-align:justify">The preprocessing step (usually an iterative process) is carried out to clean the data based on data quality issues identified. After assessing the quality of the data, we concluded that the data was clean and provides no opportunity for data cleaning. However, we identified which new columns to be added to our data to enable us to perform any suitable analysis.</p>
In this task we will perform all computations needed to create our dashboard.

##### Data Conversion

Here, we will convert the log annual income of borrowers to actual annual income.

In [11]:
#using the exp() function to get the actual annual income

income = np.exp(loan_data['log.annual.inc'])
loan_data['annual.income'] = income  #adds new column to that loan data

In [12]:
#inspecting the updated data
loan_data.head()

credit.policy             purpose  int.rate  installment  log.annual.inc  \
0              1  debt_consolidation    0.1189       829.10       11.350407   
1              1         credit_card    0.1071       228.22       11.082143   
2              1  debt_consolidation    0.1357       366.86       10.373491   
3              1  debt_consolidation    0.1008       162.34       11.350407   
4              1         credit_card    0.1426       102.92       11.299732   

     dti  fico  days.with.cr.line  revol.bal  revol.util  inq.last.6mths  \
0  19.48   737        5639.958333      28854        52.1               0   
1  14.29   707        2760.000000      33623        76.7               0   
2  11.63   682        4710.000000       3511        25.6               1   
3   8.10   712        2699.958333      33667        73.2               1   
4  14.97   667        4066.000000       4740        39.5               0   

   delinq.2yrs  pub.rec  not.fully.paid  annual.income  
0            0        0               0   85000.000385  
1            0        0               0   65000.000073  
2            0        0               0   31999.999943  
3            0        0               0   85000.000385  
4            1        0               0   80799.999636

##### Grouping Values by Conditions

###### **Categorizing FICO values**

In [13]:
#group fico values into various categories

length = len(loan_data)
ficocat = []
for i in range(0,length):
    value = loan_data['fico'][i]
    try:
        if value >= 300 and value < 400:
            cat = 'Very Poor'
        elif value >= 400 and value < 600:
            cat = 'Poor'
        elif value >= 601 and value < 660:
            cat = 'Fair'
        elif value >= 660 and value < 700:
            cat = 'Good'
        elif value >= 700:
            cat = 'Excellent'
        else:
            cat = 'Unknown'
    except:
        cat = 'Error - Unknown'
    ficocat.append(cat)
    
ficocat = pd.Series(ficocat)

In [14]:
#adding new column to loan data

loan_data['fico.category'] = ficocat

In [15]:
#inspecting updated data

loan_data.head()

credit.policy             purpose  int.rate  installment  log.annual.inc  \
0              1  debt_consolidation    0.1189       829.10       11.350407   
1              1         credit_card    0.1071       228.22       11.082143   
2              1  debt_consolidation    0.1357       366.86       10.373491   
3              1  debt_consolidation    0.1008       162.34       11.350407   
4              1         credit_card    0.1426       102.92       11.299732   

     dti  fico  days.with.cr.line  revol.bal  revol.util  inq.last.6mths  \
0  19.48   737        5639.958333      28854        52.1               0   
1  14.29   707        2760.000000      33623        76.7               0   
2  11.63   682        4710.000000       3511        25.6               1   
3   8.10   712        2699.958333      33667        73.2               1   
4  14.97   667        4066.000000       4740        39.5               0   

   delinq.2yrs  pub.rec  not.fully.paid  annual.income fico.category  
0            0        0               0   85000.000385     Excellent  
1            0        0               0   65000.000073     Excellent  
2            0        0               0   31999.999943          Good  
3            0        0               0   85000.000385     Excellent  
4            1        0               0   80799.999636          Good

- Now, we will create additional column (`int.rate.type`) to tell which borrower's interest rate is considered high or low.

In [16]:
#creating interest rate types

loan_data.loc[loan_data['int.rate'] > 0.12, 'int.rate.type'] = 'High'
loan_data.loc[loan_data['int.rate'] <= 0.12, 'int.rate.type'] = 'Low'

In [17]:
#inspecting updated data

loan_data.head()

credit.policy             purpose  int.rate  installment  log.annual.inc  \
0              1  debt_consolidation    0.1189       829.10       11.350407   
1              1         credit_card    0.1071       228.22       11.082143   
2              1  debt_consolidation    0.1357       366.86       10.373491   
3              1  debt_consolidation    0.1008       162.34       11.350407   
4              1         credit_card    0.1426       102.92       11.299732   

     dti  fico  days.with.cr.line  revol.bal  revol.util  inq.last.6mths  \
0  19.48   737        5639.958333      28854        52.1               0   
1  14.29   707        2760.000000      33623        76.7               0   
2  11.63   682        4710.000000       3511        25.6               1   
3   8.10   712        2699.958333      33667        73.2               1   
4  14.97   667        4066.000000       4740        39.5               0   

   delinq.2yrs  pub.rec  not.fully.paid  annual.income fico.category  \
0            0        0               0   85000.000385     Excellent   
1            0        0               0   65000.000073     Excellent   
2            0        0               0   31999.999943          Good   
3            0        0               0   85000.000385     Excellent   
4            1        0               0   80799.999636          Good   

  int.rate.type  
0           Low  
1           Low  
2          High  
3           Low  
4          High

## 3. Data Visualisation<a id='dv'></a>

[Move Up](#mu)

As stated earlier, we are going to create a to report of borrowers who may have issues paying back the loan in a Tableau Dashboard. We will export the processed data and load unto Tableau to build the dashboard.

In [18]:
#export data to csv

loan_data.to_csv('loan_Cleaned.csv', index=True)

## 4. References<a id='r'></a>

[Move Up](#mu)

* [Date File: loan_data_json.json](https://finch-groundhog-9245.squarespace.com/s/loan_data_json.json)
* [Blue Bank Logo](https://finch-groundhog-9245.squarespace.com/s/Blue-Bank-Logo.png)

Link to dashboard on Tableau: [Blue Bank Loan Analysis Dashboard](https://public.tableau.com/views/BlueBankLoanDashboard_16908686164080/Dashboard1?:language=en-US&publish=yes&:display_count=n&:origin=viz_share_link)